addons/folo/common/src/main/java/org/commonjava/indy/folo/data/FoloRecordCassandra.java

```diff
@@ -0,0 +1,433 @@
+                dtxTrackingRecord.setLocalUrl(localUrl);
+                trackingMapper.save(dtxTrackingRecord); //  optional Options with TTL, timestamp...
+                return true;
+            }
+
+        } catch (MalformedURLException e) {
+            throw new IndyWorkflowException( "Mailformed URL  for record: {}  path: {}", entry.getTrackingKey(),entry.getPath() );
+        }
+
+
+    }
+
+    @Override
+    public void delete(TrackingKey key) {
+
+        // Do we  want to remove records from DB?
+
+    }
+
+    @Override
+    public void replaceTrackingRecord(TrackedContent record) {
+
+        saveTrackedContentRecords(record);
+
+    }
+
+    @Override
+    public boolean hasRecord(TrackingKey key) {
+        return hasSealedRecord(key) || hasInProgressRecord(key);
+    }
+
+    @Override
+    public boolean hasSealedRecord(TrackingKey key) {
+
+        BoundStatement bind = getTrackingRecordsByTrackingKey.bind(key);
+        ResultSet execute = session.execute(bind);
+        Row one = execute.one();
+        if(one != null) {
+            Boolean sealed = one.getBool("sealed");
+            if(sealed) {
+                return true;
+            } else {
+                return false;
+            }
+        }else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean hasInProgressRecord(TrackingKey key) {
+        BoundStatement bind = getTrackingRecordsByTrackingKey.bind(key);
+        ResultSet execute = session.execute(bind);
+        Row one = execute.one();
+        if(one != null) {
+            Boolean sealed = one.getBool("sealed");
+            if(!sealed) {
+                return true;
+            } else {
+                return false;
+            }
+        }else {
+            return false;
+        }
+    }
+
+    @Override
+    public TrackedContent get(TrackingKey key) {
+        List<DtxTrackingRecord> trackingRecords =  getDtxTrackingRecordsFromDb(key);
+        return transformDtxTrackingRecordToTrackingContent(key,trackingRecords);
+    }
+
+    @Override
+    public TrackedContent seal(TrackingKey trackingKey) {
+        List<DtxTrackingRecord> trackingRecords =  getDtxTrackingRecordsFromDb(trackingKey);
+        DtxTrackingRecord recordCheck = trackingRecords.get(0);
+        if(recordCheck.getState()) {
+            logger.debug( "Tracking record: {} already sealed! Returning sealed record.", trackingKey );
+            return transformDtxTrackingRecordToTrackingContent(trackingKey,trackingRecords);
+        }
+        logger.debug( "Sealing record for: {}", trackingKey );
+        for(DtxTrackingRecord record : trackingRecords) {
+            record.setState(true);
+            trackingMapper.save(record);
+        }
+        return transformDtxTrackingRecordToTrackingContent(trackingKey,trackingRecords);
+    }
+
+    @Override
+    public Set<TrackingKey> getInProgressTrackingKey() {
+        BoundStatement inProgress = getTrackingRecordBySealed.bind(false);
```

Comment:
If we scan the full table, I suspect the query will be slow even it has index.
IMO, we may use a seperate table for inprogress records. Those records are ephemeral. ie. once we sealed them, they are not useful any more.

Reply:
No, based on tracking_key as partition key. queries. will be with best. performance. because. records will be grouped on nodes.

Reply:
This query seems just "SELECT * FROM records WHERE sealed=?" . This is a full table scan and partition key does not help.

Reply:
Yes, you are right. about that one, i've thinking of something else in. previous comment :) but nevertheless,  that's way i created indexing based on 'sealed',  from documentation:

"An index provides a means to access data in Cassandra using attributes other than the partition key. The benefit is fast, efficient lookup of data matching a given condition. The index indexes column values in a separate, hidden table from the one that contains the values being indexed."

So there is no loosing performance.

Reply:
I've suggested collapsing these two methods, in a comment above.

Reply:
> I've suggested collapsing these two methods, in a comment above.
yes, i agree